### PR TITLE
CHE 990: Add push to device feature

### PR DESCRIPTION
**Pull request by dimasnurenko**
@vparfonov @garagatyi

### CODE REVIEW

File: assembly/assembly-ide-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java [hunk 1]
```diff
@@ -138,6 +139,6 @@ protected void configure() {
 
         bind(org.eclipse.che.api.machine.server.terminal.MachineTerminalLauncher.class);
 
-        install(new org.eclipse.che.plugin.machine.artik.ArtikMachineModule());
+        install(new ArtikMachineModule());
```
Comment (garagatyi) [inline assembly/assembly-ide-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java]: Use FQN

File: assembly/assembly-machine-war/src/main/java/org/eclipse/che/wsagent/server/WsAgentModule.java [hunk 2]
```diff
@@ -52,6 +53,7 @@
     @Override
     protected void configure() {
         bind(ApiInfoService.class);
+        bind(ScpService.class);
```
Comment (garagatyi) [inline assembly/assembly-machine-war/src/main/java/org/eclipse/che/wsagent/server/WsAgentModule.java]: Use FQN

File: assembly/assembly-ide-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java (hunk 1 above)
Comment (dimasnurenko) [inline assembly/assembly-ide-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java]: ok

File: plugins/plugin-artik-ide/src/main/java/org/eclipse/che/plugin/artik/ide/ArtikExtension.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+@Singleton
+@Extension(title = "Artik", version = "1.0.0")
+public class ArtikExtension {
+
+    @Inject
+    public ArtikExtension(EventBus eventBus, final Provider<SecureCopyManager> scpManagerProvider) {
+        eventBus.addHandler(WsAgentStateEvent.TYPE, new WsAgentStateHandler() {
+            @Override
+            public void onWsAgentStarted(WsAgentStateEvent wsAgentStateEvent) {
+                scpManagerProvider.get();
```
Comment (garagatyi) [inline plugins/plugin-artik-ide/src/main/java/org/eclipse/che/plugin/artik/ide/ArtikExtension.java]: Can you explain what is happening here?

Comment (dimasnurenko) [inline plugins/plugin-artik-ide/src/main/java/org/eclipse/che/plugin/artik/ide/ArtikExtension.java]: I have to create instance after starting of worksapce agent for that i use provider for lazy initialization. I'll rewrite this code to make it more understandable.

File: plugins/plugin-artik-ide/src/main/java/org/eclipse/che/plugin/artik/ide/scp/PushToDeviceManager.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    @Override
+    public void onMachineCreating(MachineStateEvent event) {
+    }
+
+    @Override
+    public void onMachineRunning(MachineStateEvent event) {
+        MachineDto machine = event.getMachine();
+        if (!"ssh".equals(machine.getConfig().getType())) {
```
Comment (garagatyi) [inline plugins/plugin-artik-ide/src/main/java/org/eclipse/che/plugin/artik/ide/scp/PushToDeviceManager.java]: With such a check push to device won't work with artik-ide. How can we workaround that?

File: plugins/plugin-artik-server/src/main/java/org/eclipse/che/plugin/machine/artik/scp/PushToDeviceService.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+ * Copyright (c) 2012-2016 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.plugin.machine.artik.scp;
```
Comment (garagatyi) [inline plugins/plugin-artik-server/src/main/java/org/eclipse/che/plugin/machine/artik/scp/PushToDeviceService.java]: Why package includes artik word?

**Comment by garagatyi** [comment]
AFAIK it is temporary solution. So, please, add @Beta annotation from Guava to the API service

File: plugins/plugin-artik-ide/src/main/java/org/eclipse/che/plugin/artik/ide/scp/PushToDeviceManager.java (hunk 4 above)
Comment (dimasnurenko) [inline plugins/plugin-artik-ide/src/main/java/org/eclipse/che/plugin/artik/ide/scp/PushToDeviceManager.java]: Why it won't work. Can you explain?

**Comment by vparfonov** [comment]
ok

**Pull request closed without merging**